### ISSUE TITLE
Scale and shift in inference

### ISSUE THREAD

**Issue by dfrumkin** [intel-isl/MiDaS#28]
Hello!

Suppose during inference you get values in some interval [a,b].  Then for visualization you scale and shift them into some region, say [0, 1].  Now, there are two ways to do this: `(x - a) / (b - a)` and `(b - x) / (b - a)`.  Naturally, you choose the first way; however, I do not actually see how we are guaranteed that the scale must be positive.

Looking at your loss function: https://gist.github.com/dvdhfnr/732c26b61a0e63a0abc8a5d769dbebd0 - you just use least squares and can easily get positive or negative scale.

Since the network has to learn ordinal relationships (near vs. far), it seems intuitive that the scale would be positive for all images if it is positive for some; however, I am not sure we are guaranteed even that.  Or is it something you did during training?

Also, I see that in the paper propose using the median and mean absolute deviation instead.  So, what did you end up using?

**Comment by ranftlr** [comment]
Your intuition is correct: Without additional constraints there is no guarantee that the scale is positive. However, the sign of the scale in our experience remains consistent (and positive) during training. A stabilizing aspect is that the network is constrained to output only non-negative values.  While this still doesn't guarantee a positive scale when also having a shift, it seems to suffice in practice. If you are worried about this you could simply enforce a non-negative scale during training by applying a ReLU to it before scaling the output (MiDaS-v1 had such a safe-guard because the output wasn't constrained to be positive). 

> Also, I see that in the paper propose using the median and mean absolute deviation instead. So, what did you end up using?

MiDaS-v2 was trained using the median + MAD, MiDaS-v1 was trained using the least squares-based loss (see here for old version of the paper: https://arxiv.org/abs/1907.01341v1).

**Comment by dfrumkin** [comment]
Great, thank you very much!

**Comment by fu123456** [comment]
Hi, can you provide the source code midas_loss.py (https://gist.github.com/dvdhfnr/732c26b61a0e63a0abc8a5d769dbebd0) without using mask? The users can open and close the parameter to decide whether the mask needs to be used. Very thanks.

**Comment by fu123456** [comment]
@dfrumkin @dfrumkin
